runtime/js/test.js

```diff
@@ -0,0 +1,211 @@
+    let maybePromise;
+    try {
+      maybePromise = this.fn();
+    } catch (error) {
+      this.#recordFailure(error);
+      return;
+    }
+
+    if (typeof maybePromise?.then !== "function") {
+      // The test fn was synchronous, we are done.
+      this.#recordSuccess();
+      return;
+    }
+
+    // The test fn returned a promise, we need to wait for completion
+    return maybePromise.then(
+      () => this.#recordSuccess(),
+      (error) => this.#recordFailure(error),
+    );
+  }
+
+  #log(passed) {
+    if (globalRoot.lastReportedTest?.location.fileName !== this.location.fileName) {
+      // TODO(bajtos) Ideally, we should report paths relative to the project root
+      console.log("\n%s", this.location.fileName.split(/[\\\/]+/g).slice(-1)[0]);
+      globalRoot.lastReportedTest = this;
+    }
+
+    let duration = Date.now() - this.started;
+    console.log(
+      "  %s %s %s",
+      passed ? green("\u2714" /* check mark */) : red("\u2716" /* heavy multiplication */),
+      this.name,
+      grey(`(${duration}ms)`),
+    );
+  }
+
+  #recordFailure(error) {
+    globalRoot.result.failed.push(new TestFailure({ testCase: this, error }));
+    this.#log(false);
+  }
+
+  #recordSuccess() {
+    globalRoot.result.passed++;
+    this.#log(true);
+  }
+}
+
+/**
+ * @param {TestLocation} location
+ * @returns {string}
+ */
+function displayTestLocation({ fileName, lineNumber, columnNumber }) {
+  return `${fileName}:${lineNumber}:${columnNumber}`;
+}
+
+class TestFailure {
+  /**
+   * @param {Object} args
+   * @param {TestCase} args.testCase
+   * @param {Error} args.error
+   */
+  constructor({ testCase, error }) {
+    this.testCase = testCase;
+    this.error = error;
+  }
+
+  render() {
+    let location = grey(`=> ${displayTestLocation(this.testCase.location)}`);
+    return `${this.testCase.name} ${location}\n${boldRed("error:")} ${
+      this.error.stack || this.error
+    }`;
+  }
+}
+
+function runNextTest() {
+  const nextTest = globalRoot.pendingTests.shift();
+
+  if (nextTest) {
+    nextTest.execute(nextTest).then(runNextTest);
+    return;
+  }
+
+  reportTestResults();
+}
+
+function reportTestResults() {
+  let duration = Date.now() - globalRoot.started;
+  let failed = globalRoot.result.failed.length;
+  let passed = globalRoot.result.passed;
+
+  if (failed) {
+    // extra spaces are intentional to show red background
+    console.error("\n%s", whiteOnRed(" FAILURES "));
+    for (const failure of globalRoot.result.failed) {
+      console.error("\n%s", failure.render());
+    }
+  }
+
+  console.log(
+    "\n%s | %s passed | %s failed %s",
+    failed ? red("FAIL") : green("ok"),
+    passed,
+    failed,
+    grey(`(${duration}ms)`),
+  );
+  // Signal test failure by creating an unhandled error,
+  // so that `zinnia` CLI returns a non-zero exit code.
+  // TODO(bajtos) Find a more elegant solution, e.g. add es areI for setting the exit code
+  if (failed) {
+    setTimeout(() => {
```

Comment:
```suggestion
    Promise.resolve().then(() => {
```

Would this be more consistent with the rest of the lib?

Reply:
Not really. The message printed for unhandled rejections is different from the message printed for uncaught errors.

My version prints this:

```
error: Uncaught [some test(s) failed]
```


Your suggestion leads to this:

```
error: Uncaught (in promise) [some test(s) failed]
```